src/ralph/backends/database/clickhouse.py

```diff
@@ -0,0 +1,358 @@
+            host (str): ClickHouse server host to connect to.
+            port (int): ClickHouse server port to connect to.
+            database (str): ClickHouse database to connect to.
+            event_table_name (str): Table where events live.
+            client_options (dict): A dictionary of valid options for the ClickHouse
+                client connection.
+        """
+        if client_options is None:
+            client_options = {
+                "date_time_input_format": "best_effort",  # Allows RFC dates
+                "allow_experimental_object_type": 1,  # Allows JSON data type
+            }
+
+        self.host = host
+        self.port = port
+        self.database = database
+        self.event_table_name = event_table_name
+
+        self.client = clickhouse_connect.get_client(
+            host=self.host,
+            port=self.port,
+            database=self.database,
+            settings=client_options,
+        )
+
+    def status(self) -> DatabaseStatus:
+        """Checks ClickHouse connection status."""
+        try:
+            self.client.query("SELECT 1")
+        except ClickHouseError:
+            return DatabaseStatus.AWAY
+
+        return DatabaseStatus.OK
+
+    @enforce_query_checks
+    def get(self, query: ClickHouseQuery = None, chunk_size: int = 500):
+        """Gets table rows and yields them."""
+        fields = ",".join(query.return_fields) if query.return_fields else "event"
+
+        sql = f"SELECT {fields} FROM {self.event_table_name}"
+
+        if query.where_clause:
+            sql += f"  WHERE {query.where_clause}"
+
+        result = self.client.query(sql).named_results()
+
+        for statement in result:
+            yield statement
+
+    @staticmethod
+    def _validate_and_parse(statement: dict) -> Tuple[uuid.UUID, datetime.datetime]:
+        """Validate and parse required statement fields."""
+        timestamp_parser = isoparser(sep="T")
+
+        try:
+            # Validate that the statement id is a UUID
+            event_id = uuid.UUID(statement["id"], version=4)
+
+            # Convert to a timestamp object that ClickHouse can accept
+            emission_time = timestamp_parser.isoparse(statement["timestamp"])
+
+            # If the emission time didn't include a timezone, set to UTC. This should
+            # be configurable?
+            if (
+                emission_time.tzinfo is None
+                or emission_time.tzinfo.utcoffset(emission_time) is None
+            ):
+                emission_time = emission_time.replace(tzinfo=pytz.utc)
+
+            return event_id, emission_time
+        except KeyError as err:
+            raise BadFormatException(
+                f"Statement is missing id or timestamp field: " f"{statement}"
+            ) from err
+        except ValueError as err:
+            raise BadFormatException(
+                f"Statement has an invalid 'id' or 'timestamp' field: " f"{statement}"
+            ) from err
```

Comment:
Why not defining a two-fields Pydantic model (`id`, `timestamp`) here and let it do the hard work (parsing + validation)?

Reply:
I've taken a stab at this in the latest commit. I'm not sure what it gets us, so perhaps I didn't do it correctly. :)

Reply:
Ok, I think it works without the custom validators which makes more sense. 👍